datafusion/core/src/dataframe/mod.rs

```diff
@@ -1803,6 +1803,17 @@ mod tests {
         Ok(())
     }
 
+    #[tokio::test]
+    async fn test_coalesce_schema() -> Result<()> {
+        let ctx = SessionContext::new();
+
+        let query = r#"SELECT COALESCE(null, 5)"#;
```

Comment:
I tried a variation of this test and found another issue:

```
let query = r#"SELECT COALESCE(column1, column2) FROM VALUES (null, 1.2)"#;
```

failed with:

```
Left:  Schema { fields: [Field { name: "coalesce(column1,column2)", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }
Right: Schema { fields: [Field { name: "coalesce(column1,column2)", data_type: Float64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }
```

Reply:
Thanks! I trace the source of this error to:
https://github.com/apache/datafusion/blob/6bbad7eda15cd1adcfb85182f6a20df6b5035842/datafusion/expr/src/logical_plan/builder.rs#L217
The choice of the DataType::Utf8 as a default value seems off, even more so when the only case for the `Option` to be `None` is if all the values were null.

I added this case as a test, as well as some others.